R/bm-tpc-h.R

```diff
@@ -0,0 +1,206 @@
+        # This view is the most accurate comparison to Arrow, however it will
+        # penalize duckdb since AFAICT `parquet_scan` is not parallelized and
+        # ends up being the bottleneck
+        if (format == "parquet") {
+          sql_query <- paste0("CREATE VIEW ", name, " AS SELECT * FROM parquet_scan('", file, "');")
+        } else if (format == "native") {
+          sql_query <- paste0("CREATE TABLE ", name, " AS SELECT * FROM parquet_scan('", file, "');")
+        }
+
+        DBI::dbExecute(con, sql_query)
+      }
+    }
+
+    # put the necessary variables into a BenchmarkEnvironment to be used when the
+    # benchmark is running.
+    BenchEnvironment(
+      # get the correct read function for the input format
+      input_func = input_functions[[engine]],
+      tpch_files = tpch_files,
+      query = tpc_h_queries[[as.character(query_num)]],
+      engine = engine,
+      con = con,
+      scale = scale,
+      query_num = query_num
+    )
+  },
+  # delete the results before each iteration
+  before_each = {
+    result <- NULL
+  },
+  # the benchmark to run
+  run = {
+    result <- query(input_func)
+  },
+  # after each iteration, check the dimensions and delete the results
+  after_each = {
+    # If the scale is < 1, duckdb has the answer
+    if (scale %in% c(0.01, 0.1, 1)) {
+      # get answers
+      answer_psv <- DBI::dbGetQuery(
+        con,
+        paste0(
+          "SELECT answer FROM tpch_answers() WHERE scale_factor = ",
+          scale,
+          " AND query_nr = ",
+          query_num,
+          ";"
+        )
+      )
+      answer <- read.delim(textConnection(answer_psv$answer), sep = "|")
+
+      if (!isTRUE(all.equal(result, answer, check.attributes = FALSE, tolerance = 0.001))) {
+        warning("\nExpected:\n", paste0(capture.output(print(answer)), collapse = "\n"))
+        warning("\n\nGot:\n", paste0(capture.output(print(result, width = Inf)), collapse = "\n"))
+        stop("The answer does not match")
+      }
+    } else {
+      # grab scale factor 1 to check dimensions
+      answer_psv <- DBI::dbGetQuery(
+        con,
+        paste0("SELECT answer FROM tpch_answers() WHERE scale_factor = 1 AND query_nr = ", query_num, ";")
+      )
+      answer <- read.delim(textConnection(answer_psv$answer), sep = "|")
+    }
+
+    # TODO: other generic validations for SF < 1?
+    stopifnot(
+      "Dims do not match the answer" = identical(dim(result), dim(answer))
+    )
+
+    result <- NULL
+  },
+  # validate that the parameters given are compatible
+  valid_params = function(params) {
+    drop <- params$engine == "duckdb" & params$format == "feather"
+    params[!drop,]
+  },
+  # packages used when specific formats are used
+  packages_used = function(params) {
+    c(params$engine, "dplyr")
+  }
+)
+
+# all queries take an input_func which is a function that will return a dplyr tbl
+# referencing the table needed.
+#' @export
+tpc_h_queries <- list(
```

Comment:
You could also do this as:

```r
tpc_h_queries <- list()
tpc_h_queries[[1]] <- function(input_func) ...
tpc_h_queries[[6]] <- function(input_func) ...
```

Could be more intuitive than stringified number names

Reply:
Ah, yes of course. This is much better